src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -496,182 +1249,315 @@ bool emberAfThermostatClusterSetActivePresetRequestCallback(
+            // Find the number of presets in the pending preset list that match the preset handle. If there are duplicate
+            // entries, return CONSTRAINT_ERROR.
+            uint8_t count = CountPresetsInPendingListWithPresetHandle(delegate, pendingPreset.GetPresetHandle().Value());
+            if (count > 1)
+            {
+                return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::ConstraintError);
+            }
+        }
 
-    if (HeatingSetpoint > MaxHeatSetpointLimit)
-        HeatingSetpoint = MaxHeatSetpointLimit;
+        // If the preset is found in the  presets attribute list and the preset is builtIn in the pending presets list
+        //  but not in the presets attribute list, return UNSUPPORTED_ACCESS.
+        if (foundMatchingPresetInPresets && (IsBuiltIn(pendingPreset) && !IsBuiltIn(matchingPreset)))
+        {
+            return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::UnsupportedAccess);
+        }
 
-    return HeatingSetpoint;
-}
+        // If the preset is found in the  presets attribute list and the preset is builtIn in the presets attribute
+        //  but not in the pending presets list, return UNSUPPORTED_ACCESS.
+        if (foundMatchingPresetInPresets && (!IsBuiltIn(pendingPreset) && IsBuiltIn(matchingPreset)))
+        {
+            return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::UnsupportedAccess);
+        }
 
-int16_t EnforceCoolingSetpointLimits(int16_t CoolingSetpoint, EndpointId endpoint)
-{
-    // Optional Mfg supplied limits
-    int16_t AbsMinCoolSetpointLimit = kDefaultAbsMinCoolSetpointLimit;
-    int16_t AbsMaxCoolSetpointLimit = kDefaultAbsMaxCoolSetpointLimit;
+        // If the presetScenario is not found in the preset types, return CONSTRAINT_ERROR.
+        PresetScenarioEnum presetScenario = pendingPreset.GetPresetScenario();
+        if (!FindPresetScenarioInPresetTypes(delegate, presetScenario))
+        {
+            return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::ConstraintError);
+        }
 
-    // Optional User supplied limits
-    int16_t MinCoolSetpointLimit = kDefaultMinCoolSetpointLimit;
-    int16_t MaxCoolSetpointLimit = kDefaultMaxCoolSetpointLimit;
+        // If the preset type for the preset scenario does not supports names and a name is specified, return CONSTRAINT_ERROR.
+        if (!PresetTypeSupportsNames(delegate, presetScenario) && pendingPreset.GetName().HasValue())
+        {
+            return SendResponseAndCleanUp(delegate, endpoint, commandObj, commandPath, imcode::ConstraintError);
+        }
 
-    // Attempt to read the setpoint limits
-    // Absmin/max are manufacturer limits
-    // min/max are user imposed min/max
+        // Enforce the Setpoint Limits for both the cooling and heating
+        Optional<int16_t> coolingSetpointValue = preset.GetCoolingSetpoint();
```

Comment:
Shouldn't this be `pendingPreset`?  Better scoping around `preset` would have avoided this.... but was this tested?

Reply:
yes. fixed. the enforcing of cooling set point was not tested. Just checked now. i see the setpoints set correctly as expected with the fix.